### PR TITLE
Implement a URL sanitizer to check for invalid URLs and Search or add…

### LINKED ISSUE #33: `http://` should be automatically prepended when the URL is laking one

### PR DESCRIPTION
**Pull request by jonathanKingston**
… a protocol otherwise. Fixes #33

### CODE REVIEW

**Comment by jonathanKingston** [comment]
Typing 'scripting.com' will take you to 'https://scripting.com' then to 'http://scripting.com' when the URL fails. This however should not (and should never) downgrade standard navigation within the tab in this manner.

File: lib/hope-tab.js [hunk 1]
```diff
@@ -58,6 +59,47 @@ window.HopeTab = fxosComponent.register('hope-tab', {
     }
   },
 
+  onError(e) {
+    debug('on error', e, this.els.input.value, this.userSubmittedUrl);
+    var networkErrors = [
+      'dnsNotFound',
+      'connectionFailure'
```
Comment (etiennesegonzac) [inline lib/hope-tab.js]: looks like we need to downgrade for `certerror` too

Comment (jonathanKingston) [inline lib/hope-tab.js]: Yeah I was a little worried by https downgrade on navigation changes within
the tab. Assuming we can prove this isn't possible I'm happy to add this in.

File: lib/hope-tab.js [hunk 2]
```diff
@@ -58,6 +59,47 @@ window.HopeTab = fxosComponent.register('hope-tab', {
...
+    };
+    if (!this.userSubmittedUrl) {
+      return;
+    }
+    var currentUrl = this.sanitizeUrl(this.els.input.value);
+    var typedUrl = this.sanitizeUrl(this.userSubmittedUrl);
+    var typedUrlObject = new URL(typedUrl);
+
+    // Has been sanitized
+    if (this.userSubmittedUrl !== typedUrl) {
```
Comment (etiennesegonzac) [inline lib/hope-tab.js]: nit: early return

``` js
if (this.userSubmittedUrl === typedUrl) return;
// rest of the code with 1 level of nesting removed :)
```

**Comment by etiennesegonzac** [comment]
That's a huuuge usability win, thanks @jonathanKingston !
There's 2 small comments to address but we can land it afterwards, we'll wait for #36 to add tests :)

File: lib/hope-tab.js (hunk 1 above)
Comment (etiennesegonzac) [inline lib/hope-tab.js]: Can we reset `userSubmittedUrl` to null in the `onLocationChange` handler to guarantee this?

**Comment by jonathanKingston** [comment]
@etiennesegonzac thanks for checking this over.
Yup wasn't sure if the location change fired on user initiated changes thanks for prompt
Should be ready for your review again.

**Comment by etiennesegonzac** [comment]
\o/

**Pull request merged**